### PR TITLE
Linux Compatibility

### LINKED ISSUE #30: Porting to linux
Hi,

Any plans on making the setup compatible with linux ?

### PR DESCRIPTION
**Pull request by nicknisi**
This adds better support for linux to the install script and installed tools. More specifically, this adds a Dockerfile to start up a quick Linux testing environment, and updates the install scripts to be more cross-compatible.

Fixes #30

### CODE REVIEW

**Comment by jrock2004** [comment]
Adding a docker file to do testing is absolutely awesome idea. To test I was creating VM to test, but using docker is fantastic idea

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
+FROM ubuntu
+
+# Update the apt repository
+RUN apt update
+
+# Install dependencies
+RUN apt-get install -y build-essential
```
Comment (jrock2004) [inline Dockerfile:7]: I know performance is probably not needed but maybe to keep it to a single layer you could do

```
RUN apt-getinstall -y build-essential file zsh git sudo ruby curl vim language-pack-en

Comment (nicknisi) [reply Dockerfile]: Fixed!

File: install.sh [hunk 2]
```diff
@@ -1,73 +1,294 @@
...
+    backup)
+        backup
+        ;;
+    link)
+        link
+        ;;
+    git)
+        setup_git
+        ;;
+    brew)
```
Comment (jrock2004) [inline install.sh:267]: Should this be something else as it seems confusing with the actual brew command to install things? I notice on line 282 for all you call the homebrew function and now this

Comment (nicknisi) [reply install.sh]: Good call. I normalized the names to be prefixed with `setup_` and changed the name of this expected argument from `brew` to `homebrew` to give it a little more distinction from the command.

**Comment by kinowarrior** [comment]
This looks excellent, and very convenient timing too.

Just finishing a working Ubuntu setup for dotfiles based off this repo - and hope to be complete after this weekend (work has been taking all my time of late, and had to put on hold). Really appreciate seeing this PR, and will def. cross-compare what I've done against this one. In fact, might even use this as a base, and then just reapply all the s/w, vim plugin & keyboard shortcut changes. Found using brew on Ubuntu helpful for a lot of the dependencies.

One thing that I'm currently struggling with is the **best option for copy/paste with tmux/vim on Ubuntu**... 
Have a really pleasant setup on MacOS but it's not portable to Ubuntu as-is. So that's something to research at the weekend.

@nicknisi - apologies, haven't checked, but does this PR address a solution for tmux copy/paste?

The other two items are:

- Using OneDark theme successfully on Ubuntu vim, but not getting the Nerdtree icons appearing correctly yet. The airline powerfonts etc work, just something wrong with the Nerdtree folder/file icons. Can you confirm this PR addresses that one?

- Finally, was also using a `tat` script from[ Thoughtbot's dotfiles](https://github.com/thoughtbot/dotfiles/blob/master/bin/tat) set up to run automatically on shell load. That's broken too atm on Ubuntu. Otherwise, just about got everything ported.

No worries if you don't have time to reply. I'll take a copy of this PR at the weekend & try running using the Docker file provided!

**Comment by nicknisi** [comment]
@kinowarrior Thanks for the feedback! I tried to answer your questions below. There are still a few quirks to work out, and I probably need to try in a Linux VM (not just a Docker container) to test the copy/paste issues. My guess is maybe Linux will need a tool like [xclip](https://github.com/astrand/xclip) installed to help handle that.

> @nicknisi - apologies, haven't checked, but does this PR address a solution for tmux copy/paste?

It should. I did some testing on macOS and it seems that reattach-to-user-namespace isn't really required with newer versions of tmux, so I removed the call to `login-shell` in the tmux.conf and tested it works on both macOS and Ubuntu.


> * Using OneDark theme successfully on Ubuntu vim, but not getting the Nerdtree icons appearing correctly yet. The airline powerfonts etc work, just something wrong with the Nerdtree folder/file icons. Can you confirm this PR addresses that one?

For these icons to work, you will need to install a patched font from [here](https://www.nerdfonts.com/font-downloads) and then set it up as the font in your terminal. This is automated on macOS with homebrew font casks but I don't think that's supported on the linux version.

> * Finally, was also using a `tat` script from[ Thoughtbot's dotfiles](https://github.com/thoughtbot/dotfiles/blob/master/bin/tat) set up to run automatically on shell load. That's broken too atm on Ubuntu. Otherwise, just about got everything ported.

I tried the `tat` script on Linux and it seems to be working for me. Alternatively, there is a [`tm`](https://github.com/nicknisi/dotfiles/blob/master/bin/tm) script that has similar functionality in this repo, which also works for me on Linux.

**Comment by jrock2004** [comment]
I was playing with the idea of using GitHub Actions to run tests that when you push changes that it would run actions to let you know if you broke something. Issue I need to figure out is running as a non-root user in GitHub Actions as it ignores `USER` in the dockerfile. Maybe this is just something I think is cool. But I could imagine setting up actions something like this:

1. Testing MacOSX
2. Testing in Ubuntu
3. Testing in WSL (Windows subsystem for Linux)
4. Testing in ARM ubuntu

But I think this is overkill, lol

**Comment by nicknisi** [comment]
I really like that if for no other reason than it's a fun excuse to play more with Docker and GitHub actions! Tests have been something I've long wanted to add, too but the idea of writing tests for bash has seemed daunting.

**Comment by jrock2004** [comment]
Well I think phase 1 can just be that it runs and does not exit 1 building and running the scripts. This was my run, but wanted to pause for now https://github.com/jrock2004/dotfiles/runs/808388508?check_suite_focus=true

**Pull request merged**